crates/xilem_html/src/app.rs

```diff
@@ -0,0 +1,122 @@
+// Copyright 2023 the Druid Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{cell::RefCell, rc::Rc};
+
+use crate::{
+    context::Cx,
+    view::{DomNode, View},
+    Message,
+};
+use xilem_core::{Id, MessageResult};
+
+pub struct App<T, V: View<T>, F: FnMut(&mut T) -> V>(Rc<RefCell<AppInner<T, V, F>>>);
+
+struct AppInner<T, V: View<T>, F: FnMut(&mut T) -> V> {
+    data: T,
+    app_logic: F,
+    view: Option<V>,
+    id: Option<Id>,
+    state: Option<V::State>,
+    element: Option<V::Element>,
+    cx: Cx,
+}
+
+pub(crate) trait AppRunner {
+    fn handle_message(&self, message: Message);
+
+    fn clone_box(&self) -> Box<dyn AppRunner>;
+}
+
+impl<T: 'static, V: View<T> + 'static, F: FnMut(&mut T) -> V + 'static> Clone for App<T, V, F> {
+    fn clone(&self) -> Self {
+        App(self.0.clone())
+    }
+}
+
+impl<T: 'static, V: View<T> + 'static, F: FnMut(&mut T) -> V + 'static> App<T, V, F> {
+    pub fn new(data: T, app_logic: F) -> Self {
+        let inner = AppInner::new(data, app_logic);
+        let app = App(Rc::new(RefCell::new(inner)));
+        app.0.borrow_mut().cx.set_runner(app.clone());
+        app
+    }
+
+    pub fn run(self, root: &web_sys::HtmlElement) {
+        self.0.borrow_mut().ensure_app(root);
+        // Latter may not be necessary, we have an rc loop.
+        std::mem::forget(self)
+    }
+}
+
+impl<T, V: View<T>, F: FnMut(&mut T) -> V> AppInner<T, V, F> {
```

Comment:
The top-level app expects the Action type to be `()`. It is expected that if you have a view with a different action type, you'll wrap it in an `Adapt` and handle the action somehow, e.g. ignoring it.